invenio_vocabularies/contrib/names/datastreams.py

```diff
@@ -76,7 +79,49 @@ def apply(self, stream_entry, **kwargs):
         except Exception:
             pass
 
-        return StreamEntry(entry)
+        stream_entry.entry = entry
+        return stream_entry
+
+
+class NamesServiceWriter(ServiceWriter):
+    """Names service writer."""
+
+    def __init__(self, *args, scheme_id="orcid", **kwargs):
+        """Constructor."""
+        self._scheme_id = scheme_id
+        super().__init__(*args, **kwargs)
+
+    def _entry_id(self, entry):
+        """Get the id from an entry."""
+        for identifier in entry.get("identifiers"):
+            if identifier.get("scheme") == self._scheme_id:
+                return identifier["identifier"]
+
+    def _resolve(self, id_):
+        """Resolve an entry given an id."""
+        return self._service.resolve(
+            self._identity, id_=id_, id_type=self._scheme_id
+        )
+
+    def write(self, stream_entry, *args, **kwargs):
```

Comment:
Maybe add a comment (or issue to shelve) in order to remember that we have "reversed" the order here because of the PID handling issues (i.e. try to update first and then create).